p4-farm-nft/src/vault.rs

```diff
@@ -0,0 +1,299 @@
+        v.staked[nft_contract_i].push(token.clone());
+        // update total staked info about this token
+        self.total_stake[nft_contract_i] += 1;
+
+        self._recompute_stake(&mut v);
+        self.vaults.insert(previous_owner_id, &v);
+        log!(
+            "Staked @{} from {}, stake_unit: {}, cheddar_rate: {}", 
+            token.clone(), nft_contract_id, v.min_stake, self.cheddar_rate
+        );
+
+        true
+    }
+    /// Returns boost stake operation status.
+    /// Stake works only for 1 NFT token coming at the moment.
+    /// Revert transfer if nft_contract (`predecessor_account_id`) not in `Contract.boost_nft_contracts`
+    pub (crate) fn internal_boost_stake(
+        &mut self,
+        previous_owner_id: &AccountId,
+        nft_contract_id: &NftContractId,
+        token: TokenId
+    ) -> bool {
+        // find index for boost token into Contract.boost_nft_contracts
+        // Option to refund if `nft_contract_i` not in required for stake NFT contracts
+        let nft_contract_i = find_acc_idx(&nft_contract_id, &self.boost_nft_contracts);
+        let mut v = self.get_vault(&previous_owner_id);
+
+        if !v.boost_nft.is_empty() {
+            log!("Account already has boost NFT deposited. You can only deposit one");
+            return false
+        }
+        log!("Staking {} NFT - you will obtain a special farming boost", nft_contract_id);
+        self.ping_all(&mut v);
+
+        let contract_token_id:ContractNftTokenId = format!("{}{}{}", nft_contract_id, NFT_DELIMETER, token);
+
+        v.boost_nft = contract_token_id.clone();
+
+        // update total staked info about this token
+        self.total_boost[nft_contract_i] += 1;
+
+        self._recompute_stake(&mut v);
+        self.vaults.insert(&previous_owner_id, &v);
+        log!(
+            "Added boost to user @{} with {}", 
+            previous_owner_id, contract_token_id.clone()
+        );
+        true
+    }
+
+    /// Returns remaining tokens from `nft_contract_id` which user has staked after function call.
+    /// Panics if `token` not find in `Vault.staked[nft_contract_i]`.
+    /// If token not declared - unstake all tokens for this `nft_contract_id`
+    pub(crate) fn internal_nft_unstake(
+        &mut self,
+        receiver_id: &AccountId,
+        nft_contract_id: &NftContractId,
+        token_id: TokenId,
+    ) -> Vec<String> {
+        // getting contract, token and user vault
+        let nft_contract_i = find_acc_idx(nft_contract_id, &self.stake_nft_tokens);
+        let mut v = self.get_vault(receiver_id);
+        let token_i = find_token_idx(&token_id, &v.staked[nft_contract_i]);
+
+        assert!(v.staked[nft_contract_i].contains(&token_id), "{}", ERR30_NOT_ENOUGH_STAKE);
+
+        // check if we are withdraw last staked token
+        if v.get_number_of_staked_tokens() == 1 {
+            log!("unstaked last staked token - closing account");
+            self.close();
+            return vec![];
```

Comment:
you can't return without updating totals

Reply:
Contract total updates in `transfer_staked_cheddar`. On close account we don’t need update vault total, we just keep this amount to recover if promise failed, but we are close account and since that we are don’t need this total